pkg/util/active_groups_test.go

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package util
+
+import (
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/atomic"
+	"golang.org/x/exp/slices"
+)
+
+const maxGroupsPerUser = 100
+
+func TestActiveGroups(t *testing.T) {
+	ag := NewActiveGroups(maxGroupsPerUser)
+
+	ag.UpdateGroupTimestampForUser("user1", "group1", time.Unix(0, 10))
+	ag.UpdateGroupTimestampForUser("user1", "group2", time.Unix(0, 15))
+	ag.UpdateGroupTimestampForUser("user1", "group3", time.Unix(0, 20))
+
+	ag.UpdateGroupTimestampForUser("user2", "group4", time.Unix(0, 10))
+	ag.UpdateGroupTimestampForUser("user2", "group5", time.Unix(0, 15))
+	ag.UpdateGroupTimestampForUser("user2", "group6", time.Unix(0, 20))
+
+	require.Nil(t, ag.PurgeInactiveGroupsForUser("user1", 5))
+	inactiveGroupsForUser1 := ag.PurgeInactiveGroupsForUser("user1", 16)
+	require.Equal(t, 2, len(inactiveGroupsForUser1))
+	require.True(t, slices.Contains(inactiveGroupsForUser1, "group1"))
+	require.True(t, slices.Contains(inactiveGroupsForUser1, "group2"))
+	require.False(t, slices.Contains(inactiveGroupsForUser1, "group3"))
```

Comment:
[nit] Just use `require.ElementsMatch()` instead of a list of asserts. The expected list is the list of groups you expect in output. `require.ElementsMatch()` doesn't require the two slices to be sorted to correctly compare them.

_Same comment applies below._